### PR TITLE
Migrate from `LegacyColor` to `bevy_color::Color`

### LINKED ISSUE #12056: bevy_color migration plan
This ticket outlines a plan for integrating bevy_color into the rest of Bevy. The plan consists of several phases.

## Phase 1

In Phase 1, the public-facing Bevy APIs will not be changed, but the bevy_color crate will be integrated in ways that do not impact end users. During this phase, the following tasks will be performed:

* Refine and enhance the bevy_color crate with additional features, such as fluent builder methods.
* Do a mass renaming of the existing `Color` type to `LegacyColor`, and then provide a type alias `Color = LegacyColor`. The purpose of this change is to allow measurement of the migration progress by searching for the symbol `LegacyColor`. Because of the type alias, no end-user code will be broken. This step should be performed by an SME so that it can be done atomically, minimizing merge conflicts with any in-flight PRs that might be impacted.
* Replace internal uses of `LegacyColor` that are not exposed to end users, especially constant color values (I counted 601 color constants in the current code base).
* Color conversion functions in bevy_render will be moved to bevy_color. What we want to end up with is a one-way dependency such that render depends on color, and not the other way around.
* We may also want to move the `From` implementations for `LegacyColor` into the same module as `LegacyColor`.

## Phase 2

In Phase 2, we will modify Bevy APIs, but only the ones that can be modified in a backwards-compatible way. For example, the Gizmo APIs which accept a `LegacyColor` argument can be changed to accept `impl Into<LinearRgba>`, allowing either the old or new color types to be passed in without needing to call `.into()`.

Also during this phase, we'll add new APIs for bundle types such as `Fog` or `BackgroundColor` which currently accept a `LegacyColor` as a property - the new APIs will have constructors that take an `impl Into<LinearRgba>`. This means that users who want to use the new color types can do so in a forward-compatible way.

## Phase 3

In Phase 3, the remaining APIs - the ones that cannot be made backwards compatible - will be migrated. This includes things like bundle properties. In most cases, low-level color attributes will use the `LinearRgba` type.

Users will need to update their code, at minimum adding a call to `.into()` when passing a legacy color into an API that accepts the new color types.

## Timing

I'm assuming that these three phases will be spread out over several Bevy release cycles. The reason for this is that we'll want to give users plenty of warning that this breaking change is coming. However, this decision - whether to try and fit all of the migration into a single Bevy release cycle or not - is above my pay grade.

### PR DESCRIPTION
**Pull request by alice-i-cecile**
# Objective

- As part of the migration process we need to a) see the end effect of the migration on user ergonomics b) check for serious perf regressions c) actually migrate the code
- To accomplish this, I'm going to attempt to migrate all of the remaining user-facing usages of `LegacyColor` in one PR, being careful to keep a clean commit history.
- Fixes #12056.

## Solution

I've chosen to use the polymorphic `Color` type as our standard user-facing API.

- [x] Migrate `bevy_gizmos`.
- [x] Take `impl Into<Color>` in all `bevy_gizmos` APIs
- [x] Migrate sprites
- [x] Migrate UI
- [x] Migrate `ColorMaterial`
- [x] Migrate `MaterialMesh2D`
- [x] Migrate fog
- [x] Migrate lights
- [x] Migrate StandardMaterial
- [x] Migrate wireframes
- [x] Migrate clear color
- [x] Migrate text
- [x] Migrate gltf loader
- [x] Register color types for reflection
- [x] Remove `LegacyColor`
- [x] Make sure CI passes

Incidental improvements to ease migration:

- added `Color::srgba_u8`, `Color::srgba_from_array` and friends
- added `set_alpha`, `is_fully_transparent` and `is_fully_opaque` to the `Alpha` trait
- add and immediately deprecate (lol) `Color::rgb` and friends in favor of more explicit and consistent `Color::srgb`
- standardized on white and black for most example text colors
- added vector field traits to `LinearRgba`: ~~`Add`, `Sub`, `AddAssign`, `SubAssign`,~~ `Mul<f32>` and `Div<f32>`. Multiplications and divisions do not scale alpha. `Add` and `Sub` have been cut from this PR.
- added `LinearRgba` and `Srgba` `RED/GREEN/BLUE`
- added `LinearRgba_to_f32_array` and `LinearRgba::to_u32`

## Migration Guide

Bevy's color types have changed! Wherever you used a `bevy::render::Color`, a `bevy::color::Color` is used instead.

These are quite similar! Both are enums storing a color in a specific color space (or to be more precise, using a specific color model). However, each of the different color models now has its own type.

TODO...

- `Color::rgba`, `Color::rgb`, `Color::rbga_u8`, `Color::rgb_u8`, `Color::rgb_from_array` are now `Color::srgba`, `Color::srgb`, `Color::srgba_u8`, `Color::srgb_u8` and `Color::srgb_from_array`.
- `Color::set_a` and `Color::a` is now `Color::set_alpha` and `Color::alpha`. These are part of the `Alpha` trait in `bevy_color`.
- `Color::is_fully_transparent` is now part of the `Alpha` trait in `bevy_color`
- `Color::r`, `Color::set_r`, `Color::with_r` and the equivalents for `g`, `b` `h`, `s` and `l` have been removed due to causing silent relatively expensive conversions. Convert your `Color` into the desired color space, perform your operations there, and then convert it back into a polymorphic `Color` enum.
- `Color::hex` is now `Srgba::hex`. Call `.into` or construct a `Color::Srgba` variant manually to convert it. 
- `WireframeMaterial`, `ExtractedUiNode`, `ExtractedDirectionalLight`, `ExtractedPointLight`, `ExtractedSpotLight`  and `ExtractedSprite` now store a `LinearRgba`, rather than a polymorphic `Color`
- `Color::rgb_linear` and `Color::rgba_linear` are now `Color::linear_rgb` and `Color::linear_rgba`
- The various CSS color constants are no longer stored directly on `Color`. Instead, they're defined in the `Srgba` color space, and accessed via `bevy::color::palettes::css`. Call `.into()` on them to convert them into a `Color` for quick debugging use, and consider using the much prettier `tailwind` palette for prototyping.
- The `LIME_GREEN` color has been renamed to `LIMEGREEN` to comply with the standard naming.
- Vector field arithmetic operations on `Color` (add, subtract, multiply and divide by a f32) have been removed. Instead, convert your colors into `LinearRgba` space, and perform your operations explicitly there. This is particularly relevant when working with emissive or HDR colors, whose color channel values are routinely outside of the ordinary 0 to 1 range.
- `Color::as_linear_rgba_f32` has been removed. Call `LinearRgba::to_f32_array` instead, converting if needed.
- `Color::as_linear_rgba_u32` has been removed. Call `LinearRgba::to_u32` instead, converting if needed.
- Several other color conversion methods to transform LCH or HSL colors into float arrays or `Vec` types have been removed. Please reimplement these externally or open a PR to re-add them if you found them particularly useful.
- Various methods on `Color` such as `rgb` or `hsl` to convert the color into a specific color space have been removed. Convert into `LinearRgba`, then to the color space of your choice.
- Various implicitly-converting color value methods on `Color` such as `r`, `g`, `b` or `h` have been removed. Please convert it into the color space of your choice, then check these properties.
- `Color` no longer implements `AsBindGroup`. Store a `LinearRgba` internally instead to avoid conversion costs.

### CODE REVIEW

File: crates/bevy_gizmos/src/arcs.rs
Comment (viridia) [inline crates/bevy_gizmos/src/arcs.rs:251]: I don't think this doc comment line is correct.

File: crates/bevy_gizmos/src/arrows.rs
Comment (viridia) [inline crates/bevy_gizmos/src/arrows.rs:29]: Update doc comment with correct namespace.

File: crates/bevy_gizmos/src/aabb.rs
Comment (cart) [inline crates/bevy_gizmos/src/aabb.rs:112]: I think we should consider adding friendly constructors for Color like we have currently. I think this:

```rust
Color::hsl(hue, 1.0, 0.5)
```

Reads (and writes) much better than

```rust
Hsla::hsl(hue, 1.0, 0.5).into()
```

Comment (cart) [reply crates/bevy_gizmos/src/aabb.rs:112]: This would also serve to reduce breakage provided our methods are 1:1 (and i think they probably can / should be)

Comment (cart) [reply crates/bevy_gizmos/src/aabb.rs:112]: (I'm not saying we should make "not breaking people" a priority. We should only add what makes sense)

Comment (bushrat011899) [reply crates/bevy_gizmos/src/aabb.rs:112]: I agree with this. Since we've delegated domain-specific knowledge of each space to its own type all 20 of these functions (10 spaces with constructors with/out alpha) would be of the form:

```rust
impl Color {
    pub const fn rgb(r: f32, g: f32, b: f32) -> Self {
        Self::Srgba(Srgba::rgb(r, g, b))
    }
}
```

File: crates/bevy_gizmos/src/gizmos.rs
Comment (bushrat011899) [inline crates/bevy_gizmos/src/gizmos.rs:166]: Really nice that we can explicitly say this function expects any color space, but `start` and `end` colors need to be in the same space.

File: examples/2d/bounding_2d.rs
Comment (bushrat011899) [inline examples/2d/bounding_2d.rs:3]: Really nice to have this option!

File: crates/bevy_gizmos/src/aabb.rs
Comment (bushrat011899) [inline crates/bevy_gizmos/src/aabb.rs:100]: Since this method generates a nice sequence of colors based on the `u32` representation of an `Entity`, maybe we could offer a more general `pub fn color_from_index(index: u32) -> Oklcha`? Feels like a pretty handy function for debugging purposes, and might be nice for end users to have access to a color palette that can be programmatically accessed.

File: crates/bevy_gizmos/src/arcs.rs
Comment (bushrat011899) [inline crates/bevy_gizmos/src/arcs.rs:58]: Might need to do some performance testing to confirm, but we probably need to go back over the various `From` implementations an annotate them with `#[inline]`. I'm not sure how good the Rust compiler is at inlining these tiny functions.

**Review by bushrat011899 — COMMENTED** [review]
Looking good so far!

File: crates/bevy_gizmos/src/aabb.rs
Comment (bushrat011899) [reply crates/bevy_gizmos/src/aabb.rs:112]: I agree. Since all the domain specific information about the various colour spaces are confined to their respective types, it should keep the `Color` type reasonably clean. It'll basically be 20 or so `const` functions:

```rust
impl Color {
    pub const fn rgb(r: f32, g: f32, b: f32) -> Self {
        Self::Srgba(Srgba::rgb(r, g, b)) 
    }
}
```

Comment (alice-i-cecile) [reply crates/bevy_gizmos/src/aabb.rs:112]: Yep, can do!

File: crates/bevy_gizmos/src/arcs.rs
Comment (alice-i-cecile) [reply crates/bevy_gizmos/src/arcs.rs:58]: Yeah agreed. I want to do perf testing towards the end of this process.

File: crates/bevy_gizmos/src/aabb.rs
Comment (alice-i-cecile) [reply crates/bevy_gizmos/src/aabb.rs:100]: Spun out into #12170.

File: crates/bevy_gizmos/src/aabb.rs
Comment (alice-i-cecile) [reply crates/bevy_gizmos/src/aabb.rs:112]: Split into https://github.com/bevyengine/bevy/pull/12171 :)

File: crates/bevy_gizmos/src/gizmos.rs
Comment (alice-i-cecile) [reply crates/bevy_gizmos/src/gizmos.rs:167]: Yeah, I think I like the current API for now.

**Comment by jdolandev** [comment]
Not to cause undo distraction right before a merge, but I think the `Lch` color space may be an appealing option for UI elements, as shown in [palgrad](https://github.com/okaneco/palgrad).

File: crates/bevy_pbr/src/light.rs
Comment (afonsolage) [inline crates/bevy_pbr/src/light.rs:116]:
```suggestion
            color: Color::WHITE,
```

File: crates/bevy_pbr/src/light.rs
Comment (afonsolage) [inline crates/bevy_pbr/src/light.rs:187]:
```suggestion
            color: Color::WHITE,
```

File: crates/bevy_pbr/src/fog.rs
Comment (IceSentry) [inline crates/bevy_pbr/src/fog.rs:410]: Not a blocker, but I think LinearRgba would benefit from having an `as_array()` function.

File: crates/bevy_pbr/src/lib.rs
Comment (afonsolage) [inline crates/bevy_pbr/src/lib.rs:339]:
```suggestion
                base_color: Color::DEEP_PINK,
```
Maybe use `DEEP_PINK` which is almost the same as this one?
`DEEP_PINK: Srgba = Srgba::new(1.0, 0.078, 0.576, 1.0);`

File: crates/bevy_pbr/src/fog.rs
Comment (IceSentry) [reply crates/bevy_pbr/src/fog.rs:410]: Ah, nvm, this PR does add that function, it's just not used here

File: crates/bevy_sprite/src/mesh2d/color_material.rs
Comment (afonsolage) [inline crates/bevy_sprite/src/mesh2d/color_material.rs:30]:
```suggestion
                color: Color::FUCHSIA,
```
`pub const FUCHSIA: Srgba = Srgba::new(1.0, 0.0, 1.0, 1.0);`

**Review by afonsolage — COMMENTED** [review]
Just minor suggestion to try keep consistency using constant colors whenever possible.

**Comment by ChristopherBiscardi** [comment]
The change to longer names (`set_a` -> `set_alpha`) is great for clarity.

LinearRgba being the AsBindGroup also makes it much more clear what kind of value the color value is in the shader for newer shader programmers.

I really like the idea of "palettes" and the included CSS palette instead of constants. Is the mentioned `tailwind` palette included here yet? or is the mention in the migration guide more meant to be a "go grab the tailwind colors from their website and build your own palette"? I've been taking this kind of approach in a number of my projects and could see an external "bevy_palette_tailwind" being useful if its not intended to be included here.

**Comment by alice-i-cecile** [comment]
@afonsolage: the example colors are a *mess*. I'm going to try and avoid changes to clean them up here, and defer to #12176 to handle it properly.

@ChristopherBiscardi thanks! And yeah, the tailwind palette is in #12080: just waiting on a final approval on licensing details. I can definitely see it being particularly useful for you: banging out quick but attractive teaching material is a prime use case.

**Comment by alice-i-cecile** [comment]
>  @Alice 🌹 I think the choice about which API to use should be done ASAP in this cycle. I've personally been thinking about this as a two stage process:
>
> 1. Quickly port to bevy_color with roughly 1:1 conceptual mapping and making uncontroversial choices such as using LinearSrgba in the renderer
> 2. Prioritize finishing the controversial enum Color vs Srgba vs LinearSrgba conversation

- @cart [on Discord](https://discord.com/channels/691052431525675048/743663673393938453/1212504831378464880).

I've removed the controversial label, but I would very much appreciate reviewer help to help catch mistakes made during migration.

**Comment by bushrat011899** [comment]
> Not to cause undo distraction right before a merge, but I think the `Lch` color space may be an appealing option for UI elements, as shown in [palgrad](https://github.com/okaneco/palgrad).

Going even further, we now have the [Oklch](https://oklch.com/) color space, which has the same intuitions as Lch, but with a modern formulation that has [better uniformity](https://bottosson.github.io/posts/oklab/). These properties should make it ideal for generating derived colours (darken, lighten, shift hue, etc.)

File: crates/bevy_pbr/src/pbr_material.rs
Comment (pablo-lua) [inline crates/bevy_pbr/src/pbr_material.rs:532]: is there a reason to have a into here, even if the type is the same type as used in base_color? `Color::into<Color>()`

File: crates/bevy_ui/src/render/mod.rs
Comment (pablo-lua) [inline crates/bevy_ui/src/render/mod.rs:267]: I was wondering if sometime in the future we can implement deref and derefmut for this types that are only wrappers in ui, would simplify some places

File: crates/bevy_ui/src/ui_material.rs
Comment (pablo-lua) [inline crates/bevy_ui/src/ui_material.rs:60]: I was wondering why in some places we use Color::RED and in others this.

**Review by pablo-lua — COMMENTED** [review]
I had a look on everything until the examples for this review for now, and have some questions

**Comment by rparrett** [comment]
> the example colors are a mess. I'm going to try and avoid changes to clean them up her

What level of feedback are you looking for on the examples? There are a number of places where named constants have been changed to float rgb values for seemingly no reason, or `Color::from` and `CONST.into()` are being used inconsistently within the same match statement, or new imports could be compacted, etc.

**Comment by alice-i-cecile** [comment]
> > the example colors are a mess. I'm going to try and avoid changes to clean them up her
> 
> What level of feedback are you looking for on the examples? There are a number of places where named constants have been changed to float rgb values for seemingly no reason, or `Color::from` and `CONST.into()` are being used inconsistently within the same match statement, or new imports could be compacted, etc.
> 
> Happy to skip sorting through all that if we're leaving all example cleanup as followup work.

No strong feelings, but I would definitely prefer that sort of feedback to be left in the form of a PR to my branch. It would be good to clean that up, but it doesn't seem like a particularly valuable use of time to do it in this PR.

File: crates/bevy_pbr/src/pbr_material.rs
Comment (alice-i-cecile) [reply crates/bevy_pbr/src/pbr_material.rs:532]: Yeah, CI caught this too.

File: crates/bevy_ui/src/render/mod.rs
Comment (alice-i-cecile) [reply crates/bevy_ui/src/render/mod.rs:267]: Yeah, I'm not opposed to that.

**Comment by rparrett** [comment]
I'm fixing up clippy.

File: examples/3d/animated_material.rs
Comment (bushrat011899) [inline examples/3d/animated_material.rs:56]: Nit for future PR: We should add hue rotation traits to make this cleaner.

File: examples/3d/deferred_rendering.rs
Comment (bushrat011899) [inline examples/3d/deferred_rendering.rs:196]:
```suggestion
                base_color: Color::hsl(0., 0., 0.533),
```

No need to have that gross `unwrap`

File: examples/3d/transmission.rs
Comment (bushrat011899) [inline examples/3d/transmission.rs:319]:
```suggestion
                    LinearRgba::from(ANTIQUE_WHITE).mix(LinearRgba::from(ORANGE_RED), 0.2),
```

File: examples/games/contributors.rs
Comment (bushrat011899) [inline examples/games/contributors.rs:123]: Nit for future PR: This looks gross. I think it would look nicer to declare `const: SELECTED: Hsla = Hsla::new(0., 0.3, 0.2, 0.92);` and then use `SELECTED.with_hue(hue).into()`.

**Review by bushrat011899 — APPROVED** [review]
A monster PR! Some very minor things barely worth mentioning, I think let's just merge this before the conflicts get any worse and worry about everything else as follow-ups!

File: examples/3d/transmission.rs
Comment (rparrett) [reply examples/3d/transmission.rs:319]: Missing a `&` in the suggestion.

File: examples/3d/lighting.rs
Comment (rparrett) [reply examples/3d/lighting.rs:71]: I'll fix these in my PR to this branch.

File: crates/bevy_color/src/linear_rgba.rs
Comment (alice-i-cecile) [reply crates/bevy_color/src/linear_rgba.rs:334]: I'm really not sure of the best way to handle alpha with color addition and subtraction. Our options:

1. Average: leads to fairly surprising behavior, non-associative.
2. Use the left-hand alpha: fairly surprising, non-commutative.
3. Set the alpha to 1: very surprising, associative and commutative.

File: crates/bevy_color/src/color.rs
Comment (alice-i-cecile) [reply crates/bevy_color/src/color.rs:53]: Yeah, I realized that I could do this to ease migration, and promptly started laughing due to how silly it is to add new immediately deprecated functions.

File: crates/bevy_color/src/linear_rgba.rs
Comment (pablo-lua) [reply crates/bevy_color/src/linear_rgba.rs:334]: Another possibility is to do a alpha multiplication, as to say "the alpha will blend in one another". If we have ½ + ¼, this will turn in ⅛. But I'm not sure of this behavior in a add function. I think 1 is the better.

Comment (rparrett) [reply crates/bevy_color/src/linear_rgba.rs:334]: In my non-expert opinion, adding colors doesn't make any sense and maybe we shouldn't provide this.

Comment (IQuick143) [reply crates/bevy_color/src/linear_rgba.rs:334]: Why is alpha singled out?
Couldn't we just treat it as we do with the other components (it gets scaled by * and / and gets added/subtracted by + and -).
That would make the most sense when thinking about colours as an affine (or rather convex) space.
`0.5 * BLACK` is somewhat poorly defined and doesn't represent a colour, but
`0.5 * BLACK + 0.5 * WHITE` makes perfect sense.
Generally, you can take any set of coefficients `a_i >= 0` that sum to 1 and do a convex combination `sum_i a_i * color_i`. I think that is completely reasonable behaviour. I'm not a color-space expert, but I think that all the colour space gamuts we have are convex sets? (Therefore this kind of operation creates valid colours.)

This is also relevant if we want to implement Point on colours and use color curves. The code for curves generally assumes a space closed under convex combinations, with a commutative and associative addition operation (and a scaling operation that distributes over addition and agrees with scalar multiplication (meaning that `(ab) * Point = a * (b * Point)`), at least as far as I remember those are the assumptions made.)

Comment (alice-i-cecile) [reply crates/bevy_color/src/linear_rgba.rs:334]: We definitely don't want to multiply or divide alpha: multiplying and dividing colors by floats is used for things like HDR and emissivity, and suddenly affecting alpha there will lead to subtle bugs.

In terms of addition and subtraction, one of the things I like about the averaging (and all other methods proposed): we stay bounded between 0 and 1 alpha (assuming inputs within that range). What does an alpha of *2* mean? An alpha of -1?

There is a secret 4th options of course: just panic with non 0 or 1 alpha! That seems particularly nasty though, and likely to slow down this operation.

Generally I think we should punt this question to follow-up work, and tackle it once we can see how colors play with both the curves and the `Animatable` trait.

Comment (superdump) [reply crates/bevy_color/src/linear_rgba.rs:334]: In the linear domain using physical units, adding RGB together makes sense as just "more light". What does adding alpha mean? Should it apply 'over' style alpha blending i.e. when adding colours A and B together, it would produce:
```rust
red: self.red * self.alpha + other.red * (1.0 - self.alpha),
green: self.green * self.alpha + other.green * (1.0 - self.alpha),
blue: self.blue * self.alpha + other.blue * (1.0 - self.alpha),
alpha: self.alpha + other.alpha * (1 - self.alpha),
```
?

Comment (IQuick143) [reply crates/bevy_color/src/linear_rgba.rs:334]: In the linear domain using physical units, what does alpha mean?

File: examples/games/contributors.rs
Comment (rparrett) [reply examples/games/contributors.rs:123]: I had some notes on other ideas to improve that example. I'll add this to the pile.

**Comment by rparrett** [comment]
Conditional LGTM on doc getting fixed and linear color math questions being resolved in some say.

(in-my-non-expert-opinion, I don't think we need the `Add` impl and would probably prefer methods that are explicit about what they are doing if the need actually arises. I couldn't find the spot where an add impl was totally necessary or was significantly cleaner but this PR is huge)

File: crates/bevy_color/src/linear_rgba.rs
Comment (viridia) [reply crates/bevy_color/src/linear_rgba.rs]: In general, I think we should avoid overloading math operators unless they have a clear and unambiguous meaning.

HDR is an important niche, but it's still a niche, and we've already run into the problem that the "meaning" of r, g, and b channels are different in an HDR context than they are in other contexts; common intuitions about color operations, such as clamping, don't necessarily hold. We risk creating confusion if we add a bunch of HDR-specific calculation operations that aren't useful outside of an HDR domain. These would be better done via traits that are clearly marked as being HDR related.

**Review by rparrett — APPROVED** [review]

**Review by pablo-lua — APPROVED** [review]

**Pull request merged**
